File: test-apps/test-metalrender/ViewController.m
//
//  ViewController.m
//  test-metalrender
//
//  Created by Botao Hu on 10/16/20.
//

#import "ViewController.h"
#import "Renderer.h"
#import <CoreMotion/CoreMotion.h>

@interface ViewController () <MTKViewDelegate, ARSessionDelegate>

@property (nonatomic, strong) ARSession *session;
@property (nonatomic, strong) Renderer *renderer;
@end


@interface MTKView () <RenderDestinationProvider>

@end


@implementation ViewController

CMMotionManager *_motionManager;
static const NSTimeInterval kAccelerometerUpdateInterval = 0.001;
static const NSTimeInterval kGyroUpdateInterval = 0.01;
NSOperationQueue *_queue;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _motionManager = [[CMMotionManager alloc] init];
    _queue = [[NSOperationQueue alloc] init];
    // Tune these to your performance prefs.
    // maxConcurrent set to anything >1 crashes HoloKit SDK.
    _queue.maxConcurrentOperationCount = 1;
    if ([_queue respondsToSelector:@selector(setQualityOfService:)]) {
      // Use highest quality of service.
      _queue.qualityOfService = NSQualityOfServiceUserInteractive;
    }
    
    _motionManager.accelerometerUpdateInterval = kAccelerometerUpdateInterval;
    [_motionManager
        startAccelerometerUpdatesToQueue:_queue
                             withHandler:^(CMAccelerometerData *accelerometerData,
                                           NSError *error) {
        NSLog(@"accelerometerData.timestamp: %f,  systemuptime: %f", accelerometerData.timestamp, [[NSProcessInfo processInfo] systemUptime]);
        
                             }];
    
    _motionManager.deviceMotionUpdateInterval = kGyroUpdateInterval;
    [_motionManager
        startDeviceMotionUpdatesToQueue:_queue
                            withHandler:^(CMDeviceMotion *motionData, NSError *error) {
        NSLog(@"motionData.timestamp: %f,  systemuptime: %f", motionData.timestamp, [[NSProcessInfo processInfo] systemUptime]);
                            }];
    
    // Create an ARSession
    self.session = [ARSession new];
    self.session.delegate = self;
    
    // Set the view to use the default device
    MTKView *view = (MTKView *)self.view;
    view.device = MTLCreateSystemDefaultDevice();
    view.backgroundColor = UIColor.clearColor;
    view.delegate = self;
    
    if(!view.device) {
        NSLog(@"Metal is not supported on this device");
        return;
    }
    
    // Configure the renderer to draw to the view
    self.renderer = [[Renderer alloc] initWithSession:self.session metalDevice:view.device renderDestinationProvider:view];
    NSLog(@"%@", NSStringFromCGRect(view.bounds));
    [self.renderer drawRectResized:view.bounds.size];
    
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTap:)];
    NSMutableArray *gestureRecognizers = [NSMutableArray array];
    [gestureRecognizers addObject:tapGesture];
    [gestureRecognizers addObjectsFromArray:view.gestureRecognizers];
    view.gestureRecognizers = gestureRecognizers;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    ARWorldTrackingConfiguration *configuration = [ARWorldTrackingConfiguration new];

    [self.session runWithConfiguration:configuration];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self.session pause];
}

- (void)handleTap:(UIGestureRecognizer*)gestureRecognize {
    ARFrame *currentFrame = [self.session currentFrame];
    
    // Create anchor using the camera's current position
    if (currentFrame) {
        
        // Create a transform with a translation of 0.2 meters in front of the camera
        matrix_float4x4 translation = matrix_identity_float4x4;
        translation.columns[3].z = -0.2;
        matrix_float4x4 transform = matrix_multiply(currentFrame.camera.transform, translation);
        
        // Add a new anchor to the session
        ARAnchor *anchor = [[ARAnchor alloc] initWithTransform:transform];
        [self.session addAnchor:anchor];
    }
}

#pragma mark - MTKViewDelegate

// Called whenever view changes orientation or layout is changed
- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size {
    [self.renderer drawRectResized:view.bounds.size];
}

// Called whenever the view needs to render
- (void)drawInMTKView:(nonnull MTKView *)view {
    NSLog(@"update start; systemuptime: %f", [[NSProcessInfo processInfo] systemUptime]);
    [self.renderer update];
    NSLog(@"update done; systemuptime: %f", [[NSProcessInfo processInfo] systemUptime]);
}

#pragma mark - ARSessionDelegate

- (void)session:(ARSession *)session didUpdateFrame:(ARFrame *)frame {
//    frame.camera.transform
//    frame.camera.intrinsics
    NSLog(@"frame.timestamp: %f,  systemuptime: %f", frame.timestamp, [[NSProcessInfo processInfo] systemUptime]);
    
//    [[NSProcessInfo processInfo] systemUptime]
}

- (void)session:(ARSession *)session didFailWithError:(NSError *)error {
    // Present an error message to the user
    
}

- (void)sessionWasInterrupted:(ARSession *)session {
    // Inform the user that the session has been interrupted, for example, by presenting an overlay
    
}

- (void)sessionInterruptionEnded:(ARSession *)session {
    // Reset tracking and/or remove existing anchors if consistent tracking is required
    
}

@end

File: test-headtracking3/ViewController.swift
//
//  ViewController.swift
//  test-headtracking3
//
//  Created by Yuchen on 2021/8/19.
//

import UIKit
import Metal
import MetalKit
import ARKit
import os.signpost

extension MTKView : RenderDestinationProvider {
}

class ViewController: UIViewController, MTKViewDelegate, ARSessionDelegate {
    
    var session: ARSession!
    var renderer: Renderer!
    var frameCnt: Int = 0
    var previousPresentedTime: CFTimeInterval = 0.0
    let logHandler = OSLog(subsystem: "com.holoi.xr.holokit.test-headtracking.test-headtracking2", category: .pointsOfInterest)

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set the view's delegate
        session = ARSession()
        session.delegate = self
        
        // Set the view to use the default device
        if let view = self.view as? MTKView {
            view.device = MTLCreateSystemDefaultDevice()
            view.backgroundColor = UIColor.clear
            view.delegate = self
            
            guard view.device != nil else {
                print("Metal is not supported on this device")
                return
            }
            
            // Configure the renderer to draw to the view
            renderer = Renderer(session: session, metalDevice: view.device!, renderDestination: view)
            
            renderer.drawRectResized(size: view.bounds.size)
        }
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(ViewController.handleTap(gestureRecognize:)))
        view.addGestureRecognizer(tapGesture)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        // Create a session configuration
        let configuration = ARWorldTrackingConfiguration()

        // Run the view's session
        session.run(configuration)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        // Pause the view's session
        session.pause()
    }
    
    @objc
    func handleTap(gestureRecognize: UITapGestureRecognizer) {
        // Create anchor using the camera's current position
        if let currentFrame = session.currentFrame {
            
            // Create a transform with a translation of 0.2 meters in front of the camera
            var translation = matrix_identity_float4x4
            translation.columns.3.z = -0.2
            let transform = simd_mul(currentFrame.camera.transform, translation)
            
            // Add a new anchor to the session
            let anchor = ARAnchor(transform: transform)
            session.add(anchor: anchor)
        }
    }
    
    // MARK: - MTKViewDelegate
    
    // Called whenever view changes orientation or layout is changed
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        renderer.drawRectResized(size: size)
    }
    
    // Called whenever the view needs to render
    func draw(in view: MTKView) {
      
        
        
           os_signpost(.begin, log: logHandler, name: "drawcall", "begin processing for frameCnt=%d, systemtime=%f", frameCnt, ProcessInfo.processInfo.systemUptime)
           print("[drawcall begin] current at \(ProcessInfo.processInfo.systemUptime)");

                 
           if let cd = view.currentDrawable {
               
               
               os_signpost(.begin, log: logHandler, name: "currentDrawable", "before presented frameCnt=%d, cd.drawableID=%d, systemtime=%f", frameCnt, cd.drawableID, ProcessInfo.processInfo.systemUptime)
               
               cd.addPresentedHandler({ [weak self] drawable in
                   guard let strongSelf = self else {
                       return
                   }
                   //let presentationDuration = drawable.presentedTime - strongSelf.previousPresentedTime
                   //strongSelf.frameRate = 1.0/presentationDuration
                   /* ... */
                   strongSelf.previousPresentedTime = drawable.presentedTime
                   
                   os_signpost(.end, log: strongSelf.logHandler, name: "currentDrawable", "presented frameCnt=%d, cd.drawableID=%d, systemtime=%f, presentedTime=%f",strongSelf.frameCnt, cd.drawableID, ProcessInfo.processInfo.systemUptime, drawable.presentedTime)

                   print("[presentedTime] current at \(ProcessInfo.processInfo.systemUptime)");

               })
           }
          
    
           
        renderer.update()
           
           os_signpost(.end, log: logHandler, name: "drawcall", "finished processing for frameCnt=%d", frameCnt)
           print("[drawcall end] current at \(ProcessInfo.processInfo.systemUptime)");
        
    }
    
    // MARK: - ARSessionDelegate
    
    func session(_ session: ARSession, didUpdate frame: ARFrame) {
     //   let logHandler = OSLog(subsystem: "com.holoi.xr.holokit.test-headtracking.test-headtracking2", category: .pointsOfInterest)

        frameCnt += 1
        os_signpost(.begin, log: logHandler, name: "ar session", "begin ar session for frameCnt=%d, frametime=%f, systemtime=%f", frameCnt, frame.timestamp, ProcessInfo.processInfo.systemUptime )
        print("\n");
        print("[arkit session] current time at \(ProcessInfo.processInfo.systemUptime), frame sensortime = \(frame.timestamp)");
        os_signpost(.end, log: logHandler, name: "ar session", "begin ar session for frameCnt=%d", frameCnt) 

    }
    func session(_ session: ARSession, didFailWithError error: Error) {
        // Present an error message to the user
        
        
    }
    
    func sessionWasInterrupted(_ session: ARSession) {
        // Inform the user that the session has been interrupted, for example, by presenting an overlay
        
    }
    
    func sessionInterruptionEnded(_ session: ARSession) {
        // Reset tracking and/or remove existing anchors if consistent tracking is required
        
    }
}
